File: app/src/main/java/com/worldventures/dreamtrips/wallet/service/lostcard/command/WalletLocationCommand.java
package com.worldventures.dreamtrips.wallet.service.lostcard.command;

import com.worldventures.dreamtrips.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.wallet.domain.entity.lostcard.ImmutableWalletLocation;
import com.worldventures.dreamtrips.wallet.domain.entity.lostcard.WalletCoordinates;
import com.worldventures.dreamtrips.wallet.domain.entity.lostcard.WalletLocation;
import com.worldventures.dreamtrips.wallet.domain.entity.lostcard.WalletLocationType;
import com.worldventures.dreamtrips.wallet.service.SmartCardLocationInteractor;
import com.worldventures.dreamtrips.wallet.service.lostcard.LostCardRepository;

import java.util.Calendar;
import java.util.List;

import javax.inject.Inject;

import io.techery.janet.Command;
import io.techery.janet.command.annotations.CommandAction;
import rx.Observable;

@CommandAction
public class WalletLocationCommand extends Command<WalletLocation> implements InjectableAction {

   @Inject SmartCardLocationInteractor locationInteractor;
   @Inject LostCardRepository locationRepository;
   private final WalletLocationType locationType;

   public WalletLocationCommand(WalletLocationType locationType) {
      this.locationType = locationType;
   }

   @Override
   protected void run(CommandCallback<WalletLocation> callback) throws Throwable {
      final ImmutableWalletLocation.Builder locationBuilder = ImmutableWalletLocation.builder()
            .createdAt(Calendar.getInstance().getTime())
            .type(locationType);
      observeLocationDetection()
            .flatMap(geoLocationCommand -> appendCoordinates(locationBuilder, geoLocationCommand.getResult()))
            .flatMap(this::saveLocation)
            .subscribe(callback::onSuccess, callback::onFail);
   }

   private Observable<DetectGeoLocationCommand> observeLocationDetection() {
      return locationInteractor.detectGeoLocationPipe()
            .createObservableResult(new DetectGeoLocationCommand());
   }

   private Observable<WalletLocation> appendCoordinates(ImmutableWalletLocation.Builder locationBuilder,
         WalletCoordinates walletCoordinates) {
      locationBuilder.coordinates(walletCoordinates);
      return Observable.just(locationBuilder.build());
   }

   private Observable<WalletLocation> saveLocation(WalletLocation location) {
      final List<WalletLocation> walletLocations = locationRepository.getWalletLocations();
      walletLocations.add(location);
      locationRepository.saveWalletLocations(walletLocations);
      return Observable.just(location);
   }
   
}

File: wallet/src/main/java/com/worldventures/wallet/service/lostcard/command/WalletLocationCommand.kt
package com.worldventures.wallet.service.lostcard.command

import com.worldventures.janet.injection.InjectableAction
import com.worldventures.wallet.domain.entity.lostcard.WalletLocation
import com.worldventures.wallet.domain.entity.lostcard.WalletLocationType
import com.worldventures.wallet.service.SmartCardLocationInteractor
import com.worldventures.wallet.service.beacon.WalletBeaconLogger
import com.worldventures.wallet.service.lostcard.LostCardRepository
import io.techery.janet.Command
import io.techery.janet.command.annotations.CommandAction
import rx.Observable
import java.util.Calendar
import javax.inject.Inject

@CommandAction
class WalletLocationCommand(private val locationType: WalletLocationType) : Command<WalletLocation>(), InjectableAction {

   @Inject lateinit var locationInteractor: SmartCardLocationInteractor
   @Inject lateinit var locationRepository: LostCardRepository
   @Inject lateinit var walletBeaconLogger: WalletBeaconLogger

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<WalletLocation>) {
      observeLocationDetection()
            .map { WalletLocation(createdAt = Calendar.getInstance().time, type = locationType, coordinates = it.result) }
            .flatMap { this.saveLocation(it) }
            .doOnError { throwable -> walletBeaconLogger.logBeacon("WalletLocationCommand error - %s", throwable.message) }
            .subscribe({ callback.onSuccess(it) }, { callback.onFail(it) })
   }

   private fun observeLocationDetection(): Observable<DetectGeoLocationCommand> = locationInteractor.detectGeoLocationPipe()
         .createObservableResult(DetectGeoLocationCommand())

   private fun saveLocation(location: WalletLocation): Observable<WalletLocation> {
      walletBeaconLogger.logBeacon("Save location - %s", location)
      val walletLocations = locationRepository.walletLocations
      walletLocations.add(location)
      locationRepository.saveWalletLocations(walletLocations)
      return Observable.just(location)
   }

}
